### ISSUE TITLE
SyntaxError: Named export 'useRouter' not found

### ISSUE THREAD

**Issue by igordanchenko** [vercel/next.js#48801]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #22 SMP Tue Jan 10 18:39:00 UTC 2023
Binaries:
  Node: 16.17.0
  npm: 8.15.0
  Yarn: 1.22.19
  pnpm: 7.1.0
Relevant packages:
  next: 13.3.2-canary.3
  eslint-config-next: N/A
  react: 18.2.0
  react-dom: 18.2.0
```

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/nextjs-router-esm-qoquxr?file=%2Fpages%2Findex.tsx

### To Reproduce

1. Open the sandbox: https://codesandbox.io/p/sandbox/nextjs-router-esm-qoquxr?file=%2Fpages%2Findex.tsx
2. Observe the following error on the home page:

```
Server Error
SyntaxError: Named export 'useRouter' not found. The requested module 'next/router.js' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'next/router.js';
const { useRouter } = pkg;
```

3. Downgrade `next` version to 13.3.0 and the error goes away.

### Describe the Bug

It appear that the release 13.3.1 includes certain changes to the named exports, so the following import no longer works in ESM:

```js
import { useRouter } from "next/router.js";
```

The above code produces the following error:

```
Server Error
SyntaxError: Named export 'useRouter' not found. The requested module 'next/router.js' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'next/router.js';
const { useRouter } = pkg;
```

A use-case for ESM import is a 3rd-party ESM-only component library that uses `next/router`.

### Expected Behavior

Named exports should work in ESM as they used to prior to 13.3.1.

**Comment by chentsulin** [comment]
The exports object and named exports are unanalyzable after next@13.3.1 due to the limitation of https://github.com/nodejs/cjs-module-lexer.

A similar issue in nextra: https://github.com/shuding/nextra/issues/1785
Changes between Next 13.3.0 and 13.3.1: https://www.diffchecker.com/3CguFwYr/

**Comment by chentsulin** [comment]
I believe this problem is caused by swc new module transformer which was introduced by this PR: https://github.com/swc-project/swc/pull/4758

cc @kdy1

**Comment by magic-akari** [comment]
The situation only occurs in the following circumstances:

1. It uses the swc to transpile code from `esm` to `cjs`, but imports the transpiled `cjs` from `esm`.
2. It is used in a nodejs environment, rather than a webpack-like bundler.
3. It is used in a nodejs environment, but the `importInterop: node` option is not enabled.

I am considering which aspect of the isuue is the key factor.

**Comment by magic-akari** [comment]
@kdy1 Do you believe that we should always emit the `cjs-module-lexer` hint, rather than only when `importInterop: node` is enabled?
This seems to make the output verbose and seems to be a catch-all for incorrect usage.

**Comment by kdy1** [comment]
Hmm.... Not sure. How another option (e.g. `interopAnnotation`) sounds to you, just for annotations?

**Comment by magic-akari** [comment]
The optimal solution should be to set up the correct dual package.
import esm from esm, import cjs from cjs.

**Comment by kdy1** [comment]
But well, I think this is due to a wrong usage

**Comment by kdy1** [comment]
Yeah, I'll try it first

**Comment by kdy1** [comment]
How do you think about making `importInterop: swc` emit cjs hints?

**Comment by kdy1** [comment]
@magic-akari Sadly `import esm from esm, import cjs from cjs.` is not an option...
I think we may need one more import interop type, like `full` or `any` which is fully interopable (like `importInterop: swc`) and emits cjs module hints (like `importInterop: node`)

**Comment by magic-akari** [comment]
> @magic-akari Sadly `import esm from esm, import cjs from cjs.` is not an option... I think we may need one more import interop type, like `full` or `any` which is fully interopable (like `importInterop: swc`) and emits cjs module hints (like `importInterop: node`)

Both interopAnnotation boolean flag or new interop type are OK.
I prefer the boolean flag.

**Comment by kdy1** [comment]
Then let's use a boolean flag. Will you work on it? If not, I'll do it

**Comment by magic-akari** [comment]
> Then let's use a boolean flag. Will you work on it? If not, I'll do it

Yes, please assign to me.

**Comment by kdy1** [comment]
Thank you!

**Comment by CobyPear** [comment]
I'm still having issues trying to import a properly bundled package with `esm` and `cjs` into a nextjs app. It does seem like the new compiler has broken a lot of things that work fine in a "normal" nodejs + typescript app. Is there some guidance as to how to properly bundle and publish a library for Next 13.1+? Experiencing a fair amount of pain here trying to understand why importing a module to a nextjs app is so different than importing a module to a node app.

**Comment by igordanchenko** [comment]
@CobyPear, it may be worth opening a separate ticket with specific example(s) as the issue described in this ticket was fixed in 13.3.2
